tr1/misc: create object mesh structure

lahm86 opened the pull request:
Resolves #1494.

#### Checklist

- [x] I have read the [coding conventions](https://github.com/LostArtefacts/TR1X/blob/master/CONTRIBUTING.md#coding-conventions)
- [x] I have added a changelog entry about what my pull request accomplishes, or it is an internal change

#### Description

This eliminates raw mesh data for objects and replaces it with a complete structure for better readability, and improved injection support. Similar to #1938, this only applies to TR1 currently, although the groundwork has been laid in `libtrx` for TR2. Best viewed commit-by-commit, although note that this work was done incrementally so some intermediate functions/logic were added to support the new and old approach at the same time. There is some improvement remaining, such as making common reader functions for vertices and faces between rooms and objects, but I've left a `TODO` to follow this up, so that this PR doesn't impact anything done already in #1938.

This has a wide impact so will need plenty testing. I've tried a couple of TRLEs as well as OG, but the more the better. Areas impacted:
- Collision with enemies, including shooting
- Drawing Lara and Bacon Lara
- Drawing Lara's hair, and its collision
- Drawing any other animating object
- Drawing gun flare
- Drawing shadows
- Drawing static meshes
- Drawing 3D pickups (on ground and HUD)
- Drawing skyboxes
- Save crystal and Midas touch reflections
- Twinkle effect on Lara's meshes while using Midas touch, and turning to gold
- Mesh effects such as exploding eggs and body parts
- Savegame handling of Lara's mesh pointers e.g. save with Magnums equipped or holstered, and ensure on load Lara's meshes are accurate. Saves from older TR1X versions and legacy ATI saves should continue to work.
- Lara's mesh swaps:
  - Gym outfit
  - Angry Lara when shooting
  - Lara's hands, torso and thighs depending on selected gun and status
  - Getting mauled by the t-rex
  - Turning to gold
  - Tomb of Tihocan cutscene (Lara's pistols should be holstered)

lahm86 committed:
libtrx/level: hide mesh functions fully from TR2

lahm86 committed:
tr1/level: merge mesh base and pointer reading

Raw data and pointers will be used in tandem to construct meshes, so
both loading functions have been combined into one. Added logging of
indices.

lahm86 committed:
libtrx/objects/types: introduce OBJECT_MESH definition

This defines an object mesh structure to eventually replace g_MeshBase,
and pointers will replace g_Meshes. Also adds reflection properties to
FACE types.

lahm86 committed:
libtrx/level: parse object mesh data into structures

This reads the raw object mesh data and builds the structure. The
meshes are not currently stored or utilized anywhere.

lahm86 committed:
libtrx/objects: create mesh storage

This creates an API for storing and retrieving meshes, and will replace
g_Meshes and g_MeshBase.

lahm86 committed:
tr1/output: create object mesh drawing routines

lahm86 committed:
tr1/collide: use OBJECT_MESH in sphere calculation

Converts collision sphere calculations to use the new object mesh
approach.

lahm86 committed:
tr1/misc: draw common items using OBJECT_MESH

Converts general static meshes, animating items, 3D pickups, gun flash,
the lightning emitter, and effects to use the new object mesh approach.
Changes reflection toggling approach to target the mesh structure.
Updates injections to manipulate the new mesh structure where
necessary.

lahm86 committed:
tr1/lara: draw Lara and hair using OBJECT_MESH

Converts drawing routines for Lara, hair and Bacon Lara, plus hair
sphere logic, to use the new object mesh approach. Gameflow mesh swaps
will now use the new approach as well, currently only applies to the
Tihocan cutscene.

lahm86 committed:
tr1/room_draw: draw skybox using OBJECT_MESH

Converts the skybox to use the new object mesh approach.

lahm86 committed:
tr1/level: calculate max vertices using OBJECT_MESH

This updates the calculation for reserving the maximum vertices for
output to use the new object mesh approach.

lahm86 committed:
tr1/output: refactor shadow definition and usage

This moves the global shadow object into the function scope module and
recreates the structure for passing to M_CalcObjectVertices. This also
meant changing other calls to this to pass the vertices and count
rather than the object mesh.

lahm86 committed:
tr1/output: remove legacy object output functions

Everything is now using the structured approach so the legacy functions
and macros can be safely deleted.

lahm86 committed:
tr1/vars: eliminate g_Meshes and g_MeshBase

This removes the final references to g_Meshes and g_MeshBase. We skip
over reading mesh data in level and injection reading as it is handled
while parsing.

rr- commented on src/libtrx/game/level/common.c:
```diff
@@ -2,13 +2,118 @@
+            GameBuf_Alloc(sizeof(FACE4) * mesh->num_tex_face4s, GBUF_MESHES);
+        for (int32_t i = 0; i < mesh->num_tex_face4s; i++) {
+            FACE4 *const face = &mesh->tex_face4s[i];
+            for (int j = 0; j < 4; j++) {
```
Let's stick to `int32_t`

rr- commented on src/libtrx/include/libtrx/game/level/common.h:
```diff
@@ -2,4 +2,8 @@
 
+#if TR_VERSION == 1
 void Level_ReadRoomMesh(int32_t room_num, VFILE *file);
+void Level_ReadObjectMeshes(
```
Could we consider keeping this API available but unused, or possibly move it back to TR1? So far, we've only been diverging types and embedded data, and this would set a new precedent by starting to conditionalize the APIs we offer in libtrx. I'd avoid that if we can.

lahm86 replied:
Updated now so the API is similar for 1 and 2. The only difference is TR2 has no concept of `ROOM_MESH` yet because the parsing is linked to the property on the room, which hasn't been changed over yet (so same place as we were in #1938).
We'll just need to be aware not to use the other functions yet in TR2, which I think is OK for now.

rr- commented on src/libtrx/include/libtrx/game/objects/common.h:
```diff
@@ -22,3 +22,18 @@ GAME_OBJECT_ID Object_GetCognate(
+    GAME_OBJECT_ID object1_id, GAME_OBJECT_ID object2_id, int32_t mesh_num);
+
+extern void Object_DrawMesh(int32_t mesh_idx, int32_t clip, bool interpolated);
+#endif
```
Similar concern here.

rr- commented on src/tr1/global/vars.c:
```diff
@@ -63,8 +63,6 @@ int32_t g_OverlayFlag = 0;
 
 ROOM *g_RoomInfo = NULL;
-int16_t *g_MeshBase = NULL;
-int16_t **g_Meshes = NULL;
```
Hell yes.

rr- commented on src/tr1/game/objects/common.c:
```diff
@@ -397,61 +392,46 @@ void Object_DrawUnclippedItem(ITEM *item)
-        obj_ptr += 3;
-        TOGGLE_REFLECTION_ENABLED(*obj_ptr++, enabled);
+    for (int32_t i = 0; i < mesh->num_flat_face3s; i++) {
+        (&mesh->flat_face3s[i])->enable_reflections = enabled;
```
We can just do `mesh->flat_face3s[i].enable_reflections`.

rr- commented on src/tr1/game/output.c:
```diff
@@ -846,7 +797,7 @@ void Output_DrawRoom(const ROOM_MESH *const mesh)
     const int16_t size, const BOUNDS_16 *const bounds, const ITEM *const item)
 {
-    g_ShadowInfo.vertex_count = g_Config.enable_round_shadow ? 32 : 8;
+    m_Shadow.vertex_count = g_Config.enable_round_shadow ? 32 : 8;
```
Can we just allocate it on the stack here? Eg. just `SHADOW_INFO shadow = { .vertex_count = … };`

lahm86 replied:
Done. Going from global scope to function scope feels particularly nice 🎉

lahm86 committed:
libtrx/misc: adjust guards for room and object meshes

This keeps the APIs open for TR1 and TR2, on the basis that we
shouldn't call anything yet for TR2 until later in decompilation.

rr- commented:
I played only very lightly with this, and haven't tested legacy saves. The code LGTM and I don't see any potential issues. Thank you for taking this on.

aredfan reviewed (approved):
Everything looks good. 👍
